dist/tools/ethos/start_network.sh

```diff
@@ -18,33 +26,69 @@ remove_tap() {
 
 cleanup() {
     echo "Cleaning up..."
-    remove_tap
     ip a d fd00:dead:beef::1/128 dev lo
-    kill ${UHCPD_PID}
+    kill -0 "${UHCPD_PID}" >/dev/null 2>&1 && kill -TERM "${UHCPD_PID}"
+    kill -0 "${ETHOS_PID}" >/dev/null 2>&1 && kill -TERM "${ETHOS_PID}"
+    remove_tap
     trap "" INT QUIT TERM EXIT
+    RUNNING=0
 }
 
 start_uhcpd() {
-    ${UHCPD} ${TAP} ${PREFIX} > /dev/null &
+    "${UHCPD}" "${TAP}" "${PREFIX}" > /dev/null &
     UHCPD_PID=$!
 }
 
-PORT=$1
-TAP=$2
-PREFIX=$3
-BAUDRATE=115200
-UHCPD="$(readlink -f "${ETHOS_DIR}/../uhcpd/bin")/uhcpd"
 
 [ -z "${PORT}" -o -z "${TAP}" -o -z "${PREFIX}" ] && {
     echo "usage: $0 <serial-port> <tap-device> <prefix> [baudrate]"
     exit 1
 }
 
-[ ! -z $4 ] && {
+[ -n "$4" ] && {
     BAUDRATE=$4
 }
 
+RUNNING=1
 trap "cleanup" INT QUIT TERM EXIT
 
-
-create_tap && start_uhcpd && "${ETHOS_DIR}/ethos" ${TAP} ${PORT} ${BAUDRATE}
+create_tap
+start_uhcpd
+if kill -0 "${UHCPD_PID}" >/dev/null 2>&1; then
+    echo "uhcpd running, pid ${UHCPD_PID}"
+else
+    echo "uhcpd failed to start" >&2
+    exit 1
+fi
+# Run ethos as a background job, but redirect the stdin of this script to the
+# ethos process (using <&0).
+# We need to do this to remember the ethos PID to be able to clean up if this
+# script is terminated from the outside (for example when using kill -TERM $PID),
+# or else ethos may remain running after the script has been terminated.
+# Observed when running start_network.sh from within another script and the other
+# script uses kill to tell this script to quit.
+"${ETHOS}" "${TAP}" "${PORT}" "${BAUDRATE}" <&0 &
+ETHOS_PID=$!
+if kill -0 "${ETHOS_PID}" >/dev/null 2>&1; then
+    echo "ethos running, pid ${ETHOS_PID}"
+else
+    echo "ethos failed to start" >&2
+    exit 1
+fi
+# Wait until ethos has been terminated, or we receive a signal to terminate.
+# Ethos will close if the user presses CTRL+D. We will receive a signal when the
+# user presses CTRL+C, or when using kill <thispid> externally.
+while [ "$RUNNING" = "1" ]; do
+    wait "${ETHOS_PID}"
+    STATUS=$?
+    if [ $STATUS -gt 128 ]; then
+        # wait caught some signal, continue until ethos has been terminated by
+        # the cleanup function (which sets RUNNING=0).
+        continue
+    else
+        # ethos terminated with the given exit code
```

Comment:
Can also say in the comment that in this case, code will also be cleaned up.

Reply:
Hmm in fact its not always the case. Yes, when unplugging the serial port, ethos crashes and actually cleans up here.

When killing `ethos` manually with `pkill` the script does not exit and start doing a while true loop.
When killing `uhcpd`, the script does not exit either.
I think we should trigger 'cleanup' manually in these case also, it should be safe to run `cleanup` even the two at a time.